<img src="https://www.promocionachile.cl/img/logos/uai.JPG" width="400" alt="utfsm-logo" align="left"/>

# LEC de Probabilidad y Estadística, Viña del mar.
## Clase 01
### Profesor: Fabián Ramírez

# Objetivos:
* Instalar Software R.
* Conocer los beneficios de útilizar un software estadístico.
* Leer bases de datos.
* Filtrar Bases de Datos

# Presentación:
* **Fabián Ramírez Díaz**
* Estudiante de Memorista de Ingeniería civil matemática UTFSM.
* Estudiante de Magíster en Ciencias Mención Matemática.
* Programa de intercambio CRUV en el IES de la PUCV.
* Cursos de pedagogía en Preuniversitario Santa María.
* Contacto:
    - Via mail: lecpye@gmail.com

# ¿Qué es R? 

R es un lenguaje de programación enfocado a objetos, utilizada generalmente por gente que quiere realizar un estudio estadístico con la finalidad de tomar una decisión.

<img src="https://www.maximaformacion.es/wp-content/uploads/2018/08/Logo-R-Software-300x263.png" width="100" alt="utfsm-logo" align="center"/>

Se prefiere frente a otros lenguajes puesto que:
* Es **gratis** y de codigo libre.
* Es simple en comparación a otros lenguajes.
* Gran comunidad de programadores.

# Instalación del R

La instalación de R depende de su sistema operativo (Windows, Linux o Mac), por tanto dejare el link de descarga oficial de R y desde el sitio web ustedes instalaran el R dependiendo de su sistema:
* https://cran.dcc.uchile.cl/

Por desgracia la plataforma para programar es poco amigable. Por tanto se recomienda instalar **R-studio, ¡además del R!** el cual es una plataforma para programar en R de forma mas amigable.

<img src="https://rstudio.com/wp-content/uploads/2018/10/RStudio-Logo.png" width="100" alt="utfsm-logo" align="center"/>



* https://rstudio.com/products/rstudio/download/

De la misma forma que para el R buscan el instalador adecuado para su sistema operativo.

# Rstudio Cloud

Este año daremos la alternativa de trabajar desde [RStudio Cloud](https://rstudio.cloud/) en esta página podras programar en R sin la necesidad de instalar nada, solo necesitas una buena conección a internet.


# Conociendo el R
## Bases de datos:
Antes de utilizar el programa R es revisar la base de datos. Para ello se recomienda abrir la base de datos mediante algún programa como excel o libre office y deben fijarse que:
* El nombre de las variables no tenga espacios, por ejemplo si una variable se llama *clase social* debes renombrarla a *clase_social*, es común en programación cambiar los ' ' por '.'. En todo caso el R automaticamente cambia los espacios por '.' aún asi sigue siendo de buena costumbre.
* Ver que la base de datos este completa. No tenga datos faltantes, o errores de tipeo.
* **Imporante**. La base de datos debe ser guardada en formato *csv* de esa forma facilita al programa leer las bases de datos.

El formato *.csv* es un formato de guardado en un excel, lo que hace es crear un archivo de texto que divide las filas y columnas del excel a traves de puntos y comas. En este formato el R es capaz de leer la base de datos de una forma rápida y sin errores.

Para guardar un archivo excel en formato *.csv* cuando colocas *guardar como* en la parte inferior del nombre debes buscar formato *.csv*

## Leer bases de datos:
Para leer una base de datos en formato *.csv* la base de datos debe estar en la misma carpeta que se encuentra a la derecha del RStudio tal como lo indica la siguiente imagen:

![im1](im1.png)


Luego de eso debes utilizar la función `read.csv('nombre_del_archivo_en_comillas')`. Además es usual guardar los datos en una variable que generalmente se le llama `data`. (se suele usar también `datos` o `df`).

En caso de querer entrar a una carpeta para tener mayor organización en tus codigos y trabajos deberas utilizar el comando `setwd('nombre_de_la_carpeta_en_comillas')`.

En el siguiente codigo se puede ver como se lee una base de datos:

In [4]:
data = read.csv('pokemon.csv')
head(data)

X. Name          Type.1 Type.2 HP Attack Defense Sp..Atk Sp..Def Speed
1 1  Bulbasaur     Grass  Poison 45  49     49      65      65     45   
2 2  Ivysaur       Grass  Poison 60  62     63      80      80     60   
3 3  Venusaur      Grass  Poison 80  82     83     100     100     80   
4 4  Mega Venusaur Grass  Poison 80 100    123     122     120     80   
5 5  Charmander    Fire          39  52     43      60      50     65   
6 6  Charmeleon    Fire          58  64     58      80      65     80   
  Generation Legendary
1 1          False    
2 1          False    
3 1          False    
4 1          False    
5 1          False    
6 1          False

Otra forma usual de leer bases de datos es directamente desde internet. Mediante la siguiente linea de código

In [2]:
pagina = 'https://raw.githubusercontent.com/Fabimath/LEC-PYE/master/precontrol3online.csv'
x <- url(pagina, "r")
x <- readLines(x)
data = read.csv(text = x)

La idea de estas lineas de códigos es para los estudiantes que no tiene acceso a una computadora puedan leer las bases de datos desde un programa para android o ios. Este programa es:

![Rcel.png](Rcel.png)

Puedes instalarlo en tu celular desde el siguiente [link](https://play.google.com/store/apps/details?id=com.krazeapps.rprogrammingcompiler&hl=es_419)

Despues de eso la base de datos esta guardada en la variable `data` entonces cada ves que llamemos a `data` estaremos llamando a la base de datos.

## Trabajar con la base de datos
Ahora algo muy importante es saber como obtener las filas de una base de datos (también llamado data frame). Para ello se utiliza el simbolo de peso `$` con ello podemos acceder a una columna de los datos. Por ejemplo si quiero obtener solo los nombres de los pokemones deberia utilizar `data$Name`. En el siguiente codigo realizaremos dicho proceso:

In [7]:
data$HP

[1]  45  60  80  80  39  58  78  78  78  44  59  79  79  45  50  60  40  45
 [19]  65  65  40  63  83  83  30  55  40  65  35  60  35  60  50  75  55  70
 [37]  90  46  61  81  70  95  38  73 115 140  40  75  45  60  75  35  60  60
 [55]  70  10  35  40  65  50  80  40  65  55  90  40  65  90  25  40  55  55
 [73]  70  80  90  50  65  80  40  80  40  55  80  50  65  90  95  95  25  50
 [91]  52  35  60  65  90  80 105  30  50  30  45  60  60  35  60  85  30  55
[109]  40  60  60  95  50  60  50  50  90  40  65  80 105 250  65 105 105  30
[127]  55  45  80  30  60  40  70  65  65  65  65  65  75  20  95  95 130  48
[145]  55 130  65  65  65  35  70  30  60  80  80 160  90  90  90  41  61  91
[163] 106 106 106 100  45  60  80  39  58  78  50  65  85  35  85  60 100  40
[181]  55  40  70  85  75 125  20  50  90  35  55  40  65  55  70  90  90  75
[199]  70 100  70  90  35  55  75  55  30  75  65  55  95  65  95  60  95  60
[217]  48 190  70  50  75 100  65  75  75  60  90  65  70  70  20  80  80  55
[235]  60  90  40  50  50 100  55  35  75  45  65  65  45  75  75  75  90  90
[253]  85  73  55  35  50  45  45  45  95 255  90 115 100  50  70 100 100 106
[271] 106 100  40  50  70  70  45  60  80  80  50  70 100 100  35  70  38  78
[289]  45  50  60  50  60  40  60  80  40  70  90  40  60  40  60  28  38  68
[307]  68  40  70  60  60  60  80 150  31  61   1  64  84 104  72 144  50  30
[325]  50  70  50  50  50  50  50  60  70  70  30  60  60  40  70  70  60  60
[343]  65  65  50  70 100  45  70  70 130 170  60  70  70  70  60  80  60  45
[361]  50  80  50  70  45  75  75  73  73  70  70  50 110  43  63  40  60  66
[379]  86  45  75  20  95  70  60  44  64  64  20  40  99  65  65  65  95  50
[397]  80  80  70  90 110  35  55  55 100  43  45  65  95  95  40  60  80  80
[415]  80  80  80  80  80  80  80 100 100 100 100 105 105 100  50  50  50  50
[433]  55  75  95  44  64  76  53  64  84  40  55  85  59  79  37  77  45  60
[451]  80  40  60  67  97  30  60  40  60  60  60  70  30  70  60  55  85  45
[469]  70  76 111  75  90 150  55  65  65  60 100  49  71  45  63 103  57  67
[487]  50  20 100  76  50  58  68 108 108 135  40  70  70  68 108  40  70  48
[505]  83  74  49  69  45  60  90  90  70  70 110 115 100  75  75  85  86  65
[523]  65  75 110  85  68  68  60  45  70  50  50  50  50  50  50  75  80  75
[541] 100  90  91 110 150 150 120  80 100  70 100 100 120 100  45  60  75  65
[559]  90 110  55  75  95  45  60  45  65  85  41  64  50  75  50  75  50  75
[577]  76 116  50  62  80  45  75  55  70  85  55  67  60 110 103 103  75  85
[595] 105  50  75 105 120  75  45  55  75  30  40  60  40  60  45  70  70  50
[613]  60  95  70 105 105  75  50  70  50  65  72  38  58  54  74  55  75  50
[631]  80  40  60  55  75  45  60  70  45  65 110  62  75  36  51  71  60  80
[649]  55  50  70  69 114  55 100 165  50  70  44  74  40  60  60  35  65  85
[667]  55  75  50  60  60  46  66  76  55  95  70  50  80 109  45  65  77  59
[685]  89  45  65  95  70 100  70 110  85  58  52  72  92  55  85  91  91  91
[703]  79  79  79  79 100 100  89  89 125 125 125  91  91 100 100  71  56  61
[721]  88  40  59  75  41  54  72  38  85  45  62  78  38  45  80  62  86  44
[739]  54  78  66 123  67  95  75  62  74  74  45  59  60  60  78 101  62  82
[757]  53  86  42  72  50  65  50  71  44  62  58  82  77 123  95  78  67  50
[775]  45  68  90  57  43  85  49  44  54  59  65  55  75  85  55  95  40  85
[793] 126 126 108  50  50  80  80  80

De esta forma podemos obtener una columna de la base de datos. Ahora también podriamos buscar datos que cumplan con una cierta condición. Recordemos que las condiciones en programación son:
* `==` para denotar igualdad entre 2 variables.
* `<=` para denotar que una variable es menor o igual a otra.
* `>=` para denotar que una variable es mayor o igual a otra.
* `<` para denotar que una variable es extrictamente menor que otra.
* `>` para denotar que una variable es extrictamente mayor que otra.
* `!=` para denotar que dos variables son distintas.

### Filtrar bases de datos
Imagina que en tu base de datos solo te sirven los datos que cumplan una condición particular. Para ello en R se utiliza la función `subset(data,condición)` esta te entrega una base de datos filtrada con la condición que tu impongas. Por ejemplo en nuestra base de datos buscaremos los Pokemons con más de 160 puntos de `Attack`, entonces:

In [10]:
fuego  = subset(data, Type.1 == 'Fire')

En R también existen las funciones clasicas de sumar,restar,multiplicar y dividir dadas por `+` `-` `*` `/`. En particular existen dos funciones `max` y `min` las cuales te entregan el máximo y el mínimo de un conjunto de datos. Por ejemplo si tu conjunto de datos es:
$$
\{ 1,2,2,0,3\}
$$
Entonces:

In [14]:
max(1,2,2,0,3)

[1] 3

In [15]:
min(1,2,2,0,3)

[1] 0

Combinado con la función `subset` podemos obtener el o los datos con mayor valor en una cierta variable. Por ejemplo podriamos buscar el Pokemon con mayor `Speed` de todos entonces:

In [12]:
subset(data,HP==min(HP))

X.  Name     Type.1 Type.2 HP Attack Defense Sp..Atk Sp..Def Speed
317 317 Shedinja Bug    Ghost  1  90     45      30      30      40   
    Generation Legendary
317 3          False

De esta forma obtenemos el Pokémon con mayor velocidad de todos. También podemos utilizar mas de una condición utilizando los conectores logicos **y** y **o** representados en R por `&` y `|` respectivamente. Por ejemplo podriamos buscar los pokemones con `HP` mayor que 130 **y** que `Type.1` sea `Water`.

In [17]:
subset(data,HP<50 & Type.1 == 'Grass')

X.  Name      Type.1 Type.2 HP Attack Defense Sp..Atk Sp..Def Speed
1     1 Bulbasaur Grass  Poison 45 49     49      65      65      45   
49   49 Oddish    Grass  Poison 45 50     55      75      65      30   
167 167 Chikorita Grass         45 49     65      49      65      45   
203 203 Hoppip    Grass  Flying 35 35     40      35      55      50   
207 207 Sunkern   Grass         30 30     30      30      30      30   
273 273 Treecko   Grass         40 45     35      65      55      70   
297 297 Seedot    Grass         40 40     50      30      30      30   
452 452 Budew     Grass  Poison 40 30     35      50      70      55   
468 468 Cherubi   Grass         45 35     45      62      53      35   
555 555 Snivy     Grass         45 45     55      45      55      63   
607 607 Cottonee  Grass  Fairy  40 27     60      37      50      66   
609 609 Petilil   Grass         45 35     50      70      50      30   
659 659 Ferroseed Grass  Steel  44 50     91      24      86      10   
    Generation Legendary
1   1          False    
49  1          False    
167 2          False    
203 2          False    
207 2          False    
273 3          False    
297 3          False    
452 4          False    
468 4          False    
555 5          False    
607 5          False    
609 5          False    
659 5          False